### PR TITLE
Authorization

### PR DESCRIPTION
**Pull request by polina606257**

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -10,10 +10,10 @@ function App() {
...
       <Routes>
         <Route path="/" element={<AppBar />}>
-          <Route path="/" element={<Home />} />
-          <Route path="shop" element={<Shop />} />
-          <Route path="contact" element={<Contact />} />
-          <Route path="sign_in" element={<SignIn />} />
+          <Route path="/" element={<Home key="home" />} />
+          <Route path="shop" element={<Shop key="shop" />} />
+          <Route path="contact" element={<Contact key="contact" />} />
+          <Route path="sign_in" element={<SignIn key="sign in" />} />
```
Comment (vldmri) [inline src/App.tsx:16]: you could have smth like `router.config.js`:
```
export default config = [{
  path: 'contact',
}]
```
and then use `config.map(...)`

File: src/assets/strings/strings.tsx [hunk 2]
```diff
@@ -0,0 +1,15 @@
+const strings = {
+  titleHaveAccount: "I already have an account",
+  sub_title_have_account: "Sign in with your email and password",
```
Comment (vldmri) [inline src/assets/strings/strings.tsx:3]: code styling, here it's snake-case, above is camel-case.
You could use `eslint` for that purpose.

File: src/assets/strings/strings.tsx [hunk 3]
```diff
@@ -0,0 +1,15 @@
+const strings = {
```
Comment (vldmri) [inline src/assets/strings/strings.tsx:1]: maybe not "strings" but rather "i18n" which stands for internationalization

File: src/components/Buttons.tsx [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+export const GoogleSignInButton: React.FC<{
+  text: string;
+  className?: string;
+  onClick: (e: React.FormEvent) => void;
+}> = (props) => {
+  const buttonClassName = `button ${props.className}`;
+  return (
+    <button className={buttonClassName} onClick={props.onClick}>
+      {props.text}
+    </button>
```
Comment (vldmri) [inline src/components/Buttons.tsx:26]: you've got quite a lot of boilerplate here, instead you could add `<BaseButton>` and build other buttons around it:
```
return (
    <BaseButton { ...props }>
      {props.text}
    </BaseButton>
```

File: src/components/auth/CreateAccountForm.tsx [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+  const handleEmailChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setFormInput((prevInput) => ({ ...prevInput, email: e.target.value }));
+  };
+
+  const handlePasswordChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setFormInput((prevInput) => ({ ...prevInput, password: e.target.value }));
+  };
+
+  const handleConfirmPasswordChange = (
```
Comment (vldmri) [inline src/components/auth/CreateAccountForm.tsx:37]: again, a boilerplate, you could have:
```
const handleChange = (fieldName: string) => {
    setFormInput({[fieldName]: e.target.value});
  };

File: src/components/auth/CreateAccountForm.tsx [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        ...prevErrors,
+        passwordError: strings.password_error,
+      }));
+      formIsValid = false;
+    }
+
+    if (formInput.confirmPassword !== formInput.password) {
+      setInputErrors((prevErrors) => ({
+        ...prevErrors,
+        confirmPasswordError: strings.password_confirm_error,
```
Comment (vldmri) [inline src/components/auth/CreateAccountForm.tsx:111]: also, boilerplate. you could have:
```
setInputErrors((prevErrors) => ({
        ...prevErrors,
        [fieldName]: strings[fieldName],
```

Comment (vldmri) [reply src/components/auth/CreateAccountForm.tsx:111]: also, you can use a "Strategy" design patter to use specific validator function for each `fieldName`, the config could be the following:
```
const inputsConfig: {
  email: {
    validator: /^[^\s@]+@[^\s@]+\.[^\s@]+$/,
    errorText: 'wrong email',
    hint: 'please enter your email here',
  },
}
```

File: src/components/auth/CreateAccountForm.tsx [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    password: "",
+    confirmPassword: "",
+  });
+  const [inputErrors, setInputErrors] = useState({
+    nameError: "",
+    emailError: "",
+    passwordError: "",
+    confirmPasswordError: "",
+  });
+  let formIsValid = true;
```
Comment (vldmri) [inline src/components/auth/CreateAccountForm.tsx:23]: this must be either part of state or `useRef` hook. You can have members in pure components.

File: src/components/auth/CreateAccountForm.tsx [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+          ...prevErrors,
+          emailError: strings.already_registered_email_error,
+        }));
+      });
+  };
+
+  const onSubmit = (e: React.FormEvent) => {
+    e.preventDefault();
+    findErrorsValidatingForm();
+    if (formIsValid) {
```
Comment (vldmri) [inline src/components/auth/CreateAccountForm.tsx:70]: this is not good for readability.
You could instead have ```
validateFormAndSubmitIfValid(() => {
    createUserWithMailAndPassword(...);
})
```
const validateFormAndSubmitIfValid = (submit: (): void) => {
   isValid = ???;
   setIsValid(isValid);
   if(isValid) submit();
}
```

```

File: src/components/auth/CreateAccountForm.tsx [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+      <Input
+        htmlFor="password"
+        labelName="Password"
+        id="password"
+        type="password"
+        value={formInput.password}
+        onChange={handlePasswordChange}
+        error={inputErrors.passwordError}
+      />
+      <Input
```
Comment (vldmri) [inline src/components/auth/CreateAccountForm.tsx:146]: also boilerplate, you could render it with `inputsConfig` from [this comment](https://github.com/polina606257/perfect_shopping/pull/2/files#r1229977522)

File: src/components/auth/CreateAccountForm.tsx [hunk 10]
```diff
@@ -74,67 +121,40 @@ const CreateAccountForm = () => {
...
-        id="confirm-password"
-        type="password"
-        value={formInput.confirmPassword}
-        onChange={(e) => handleChange("confirmPassword", e)}
-        error={inputErrors.confirmPasswordError}
-      />
+      {Object.entries(inputsConfig).map(([fieldName, config]) => (
+        <Input
+          key={fieldName}
+          htmlFor={config.htmlFor}
```
Comment (vldmri) [inline src/components/auth/CreateAccountForm.tsx:149]: `htmlFor={fieldName}`

File: src/components/auth/CreateAccountForm.tsx [hunk 11]
```diff
@@ -74,67 +121,40 @@ const CreateAccountForm = () => {
...
+          labelName={config.labelName}
+          id={config.id}
```
Comment (vldmri) [inline src/components/auth/CreateAccountForm.tsx:151]: `id={fieldName}`

File: src/components/auth/CreateAccountForm.tsx [hunk 12]
```diff
@@ -74,67 +121,40 @@ const CreateAccountForm = () => {
...
+          type={config.type}
+          value={config.value()}
+          onChange={config.onChange}
```
Comment (vldmri) [inline src/components/auth/CreateAccountForm.tsx:154]:
```
onChange={(e: React.ChangeEvent<HTMLInputElement>) => handleChange(fieldName, e)}
```

File: src/components/auth/CreateAccountForm.tsx [hunk 13]
```diff
@@ -74,67 +121,40 @@ const CreateAccountForm = () => {
     });
     formIsValidRef.current = true;
 
-    if (formInput.name.length < 2) {
-      handleError("nameError", t("name_error"));
-      formIsValidRef.current = false;
-    }
+    Object.entries(inputsConfig).forEach(([fieldName, config]) => {
+      const { validator, errorText } = config;
```
Comment (vldmri) [inline src/components/auth/CreateAccountForm.tsx:125]:
```
const errorText = config[fieldName].errorText;
```

Comment (vldmri) [inline src/components/auth/CreateAccountForm.tsx:125]:
```
const errorText = t(`${fieldName}-error`);
```

File: src/components/auth/CreateAccountForm.tsx [hunk 14]
```diff
@@ -30,6 +30,53 @@ const CreateAccountForm = () => {
...
+    confirmPassword: {
+      htmlFor: "confirm-password",
+      labelName: "Confirm password",
+      id: "confirm-password",
+      type: "password",
+      value: () => formInput.confirmPassword,
+      onChange: (e: React.ChangeEvent<HTMLInputElement>) =>
+        handleChange("confirmPassword", e),
+      validator: (value: string) => value === formInput.password,
+      errorText: t("password_confirm_error"),
```
Comment (vldmri) [inline src/components/auth/CreateAccountForm.tsx:76]: check [this comment](https://github.com/polina606257/perfect_shopping/pull/2/commits/e32c36b3525905264c0bf56ad5d25b5af71ff9aa#r1233970979)

File: src/components/auth/CreateAccountForm.tsx (hunk 11 above)
Comment (vldmri) [reply src/components/auth/CreateAccountForm.tsx:151]: actually, why do you need `id` attribute? you already have `name` / `key`

**Comment by polina606257** [comment]
authorization

**Pull request merged**